### DISCUSSION TITLE
Data corruption - yes or no?

### DISCUSSION THREAD

**Discussion by tslivnik** [openzfs/zfs#13630] [Q&A]
I had `zpool status` report this earlier today:

```
  pool: mntz
 state: DEGRADED
status: One or more devices are faulted in response to persistent errors.
	Sufficient replicas exist for the pool to continue functioning in a
	degraded state.
action: Replace the faulted device, or use 'zpool clear' to mark the device
	repaired.
  scan: scrub in progress since Fri Jul  1 00:00:01 2022
	31.2T scanned out of 40.1T at 65.0M/s, 39h56m to go
	1.38M repaired, 77.75% done
config:

	NAME                                    STATE     READ WRITE CKSUM
	mntz                                    DEGRADED     0     0     0
	  mirror-0                              ONLINE       0     0     0
	    ata-WDC_WD121KRYZ-01W0RB0_XXXXXXXX  ONLINE       0     0     0
	    ata-WDC_WD121KRYZ-01W0RB0_XXXXXXXX  ONLINE       0     0     0
	  mirror-1                              ONLINE       0     0     0
	    ata-WDC_WD101KRYZ-01JPDB0_XXXXXXXX  ONLINE       0     0     0
	    ata-WDC_WD101KRYZ-01JPDB0_XXXXXXXX  ONLINE       0     0     0
	  mirror-2                              DEGRADED     0     0     0
	    ata-WDC_WD121KRYZ-01W0RB0_XXXXXXXX  FAULTED      0     0     0  too many errors  (repairing)
	    ata-WDC_WD121KRYZ-01W0RB0_XXXXXXXX  ONLINE       0     0     0  (repairing)
	  mirror-3                              ONLINE       0     0     0
	    ata-WDC_WD181KRYZ-01AGBB0_XXXXXXXX  ONLINE       0     0     0
	    ata-WDC_WUH721818ALE6L4_XXXXXXXX    ONLINE       0     0     0
```
On reviewing the kernel log, I saw that both the drives in mirror-2 had experienced some I/O errors 1 hour apart from each other. In the past, whenever one drive in a mirror was FAULTED due to too many errors, I always cleared the zpool, which caused a resilver, and everything was honky dory.

This was the first time both drivers in a mirror experienced I/O errors (nearly) simultanelously - before I could catch the first error and resilver the damaged drive. Moreover, this is the first time that a drive was FAULTED in the middle of a scrub.

I ran "zpool clear" regardless, as I wanted the FAULTED drive to be brought online without delay and the ongoing scrub had 40 hours left to run. Now `zpool status` says:

```
  pool: mntz
 state: DEGRADED
status: One or more devices is currently being resilvered.  The pool will
	continue to function, possibly in a degraded state.
action: Wait for the resilver to complete.
  scan: resilver in progress since Wed Jul  6 20:10:58 2022
	70.0G scanned out of 40.1T at 71.0M/s, 164h17m to go
	572K resilvered, 0.17% done
config:

	NAME                                    STATE     READ WRITE CKSUM
	mntz                                    DEGRADED     0     0     0
	  mirror-0                              ONLINE       0     0     0
	    ata-WDC_WD121KRYZ-01W0RB0_XXXXXXXX  ONLINE       0     0     0
	    ata-WDC_WD121KRYZ-01W0RB0_XXXXXXXX  ONLINE       0     0     0
	  mirror-1                              ONLINE       0     0     0
	    ata-WDC_WD101KRYZ-01JPDB0_XXXXXXXX  ONLINE       0     0     0
	    ata-WDC_WD101KRYZ-01JPDB0_XXXXXXXX  ONLINE       0     0     0
	  mirror-2                              DEGRADED     0     0     0
	    ata-WDC_WD121KRYZ-01W0RB0_XXXXXXXX  FAULTED      0     0     0  too many errors  (resilvering)
	    ata-WDC_WD121KRYZ-01W0RB0_XXXXXXXX  ONLINE       0     0     0
	  mirror-3                              ONLINE       0     0     0
	    ata-WDC_WD181KRYZ-01AGBB0_XXXXXXXX  ONLINE       0     0     0
	    ata-WDC_WUH721818ALE6L4_XXXXXXXX    ONLINE       0     0     0

errors: No known data errors

```
Evidently, the faulted drive in the mirror is suffering further I/O errors even during the resilver.

Unfortunately, I have not kept a copy of the final status line reported by the first "zpool status".

What I'd like to know is if all these errors have caused any data corruption and if so, in which files.
Have I lost/masked this information with the "zpool clear", or does the fact that I'm getting "No known data errors" mean that my data is safe?

Does the fact that both mirrors said "(repairing)" mean that there was data corruption on both drives of the mirror vdev? Was it dumb to clear the zpool - should I have waited for the scrub to complete - despite the risks inherent in running the filesystem for 2 days with only a single ONLINE copy?

Thank you.

**Comment by rincebrain** [comment]
`zpool clear` can't generally delete notifications of permanent errors - they'd be mentioned in the `no known data errors` section, and even then, those can be incorrect about it being permanent and uncorrectable now if they don't know something about the system you do, like that it errored because you yanked the drive and the data is fine on the disk now that it's back.

If you don't see anything listed after a scrub, you're likely fine. Even if you did, it won't let you randomly read mangled data, it'll return EIO if you try.
